### PR TITLE
#1922 Make PBR scale and offset crosshair work like Texture crosshair do

### PR DESCRIPTION
**Pull request by LLGuru**

### CODE REVIEW

File: indra/llrender/llimagegl.cpp [hunk 1]
```diff
@@ -76,10 +76,18 @@ void LLImageGLMemory::alloc_tex_image(U32 width, U32 height, U32 pixformat)
...
+    if (iter == sTextureAllocs.end())
+    {
+        sTextureBytes += size;
+        sTextureAllocs.emplace(texName, size);
+    }
+    else
+    {
+        sTextureBytes += size - iter->second;
+        iter->second = size;
+    }
```
Comment (akleshchev) [inline indra/llrender/llimagegl.cpp:90]: Please remove this from the PR, this is wrong.

File: indra/newview/llface.cpp [hunk 2]
```diff
@@ -567,8 +567,17 @@ void LLFace::renderSelected(LLViewerTexture *imagep, const LLColor4& color)
         {
             // cheaters sometimes prosper...
             //
-            mVertexBuffer->setBuffer();
-            mVertexBuffer->draw(LLRender::TRIANGLES, mIndicesCount, mIndicesIndex);
+            LLVertexBuffer* vertex_buffer = mVertexBuffer.get();
```
Comment (akleshchev) [inline indra/newview/llface.cpp:570]: Please add a coment that this is needs to clean buffer from getGeometryVolume

**Comment by akleshchev** [comment]
```
/Users/runner/work/viewer/viewer/indra/newview/llface.cpp:1528:63: error: left operand of comma operator has no effect [-Werror,-Wunused-value]
                        s_scale = gltf_mat->mTextureTransform[0, LLGLTFMaterial::GLTF_TEXTURE_INFO_NORMAL].mScale[VX];
                                                              ^
/Users/runner/work/viewer/viewer/indra/newview/llface.cpp:1529:63: error: left operand of comma operator has no effect [-Werror,-Wunused-value]
                        t_scale = gltf_mat->mTextureTransform[0, LLGLTFMaterial::GLTF_TEXTURE_INFO_NORMAL].mScale[VY];
                                                              ^
2 errors generated.
```

File: indra/newview/llface.cpp [hunk 3]
```diff
@@ -1485,7 +1523,15 @@ bool LLFace::getGeometryVolume(const LLVolume& volume,
...
-                    tep->getScale( &s_scale, &t_scale );
+                    if (gltf_mat && rebuild_for_gltf)
+                    {
+                        s_scale = gltf_mat->mTextureTransform[0, LLGLTFMaterial::GLTF_TEXTURE_INFO_NORMAL].mScale[VX];
+                        t_scale = gltf_mat->mTextureTransform[0, LLGLTFMaterial::GLTF_TEXTURE_INFO_NORMAL].mScale[VY];
+                    }
+                    else
+                    {
+                        tep->getScale(&s_scale, &t_scale);
+                    }
```
[the remark this answers is not in the document]
Comment (LLGuru) [reply indra/newview/llface.cpp:1534]: > What does this do?

The inline function getScale(F32*, F32*) is just a syntax sugar retrieveing two numbers in one line
image
I simply preserved the default behavior in this `else` clause

**Review by akleshchev — CHANGES_REQUESTED** [review]
I tested it, and it doesn't work for layers/channels.
image

P.S. I assume that's what sRenderHighlightTextureChannel is for.

File: indra/newview/llface.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (LLGuru) [reply indra/newview/llface.cpp:1548]: > Yes, I get that, but how is this related to drawing crosshair?

The same way as before:
image

> And if it is actually related to croshair, I suggest adding a comment to that effect.
I wonder how it was working before without a such comment

[the remark this answers is not in the document]
Comment (akleshchev) [reply indra/newview/llface.cpp:1548]: >The same way as before:

The only thing that scale is used for is `bump_s_primary_light_ray.load3((offset_multiple * s_scale * primary_light_ray).mV);`, which has zero apparent relationship to a crosshair.

How is that related to crosshair?

[the remark this answers is not in the document]
Comment (LLGuru) [reply indra/newview/llface.cpp:1548]: You're right, it doesn't seem to relate to the selection marker

Fixed

File: indra/newview/llpanelface.cpp [hunk 4]
```diff
@@ -1132,65 +1107,51 @@ void LLPanelFace::updateUI(bool force_set_values /*false*/)
...
+            mLabelShiniColor->setEnabled(editable);
 
-        getChild<LLUICtrl>("combobox shininess")->setTentative(!identical_spec);
-        getChild<LLUICtrl>("glossiness")->setTentative(!identical_spec);
-        getChild<LLUICtrl>("environment")->setTentative(!identical_spec);
-        mShinyColorSwatch->setTentative(!identical_spec);
+            mComboShininess->setTentative(!identical_spec);
+            mGlossiness->setTentative(!identical_spec);
+            mEnvironment->setTentative(!identical_spec);
+            mShinyColorSwatch->setTentative(!identical_spec);
```
Comment (akleshchev) [inline indra/newview/llpanelface.cpp:1135]: Please move all there changes to a separate commit.

Not only they are massive and make checking your review troublesome, in case of any issues tracking changes or reverting will be hard.

**Comment by akleshchev** [comment]
```
/Users/runner/work/viewer/viewer/indra/newview/llface.cpp:1252:17: error: 4 enumeration values not handled in switch: 'DIFFUSE_MAP', 'ALTERNATE_DIFFUSE_MAP', 'SPECULAR_MAP'... [-Werror,-Wswitch]
        switch (LLPipeline::sRenderHighlightTextureChannel)
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/runner/work/viewer/viewer/indra/newview/llface.cpp:1252:17: note: add missing switch cases
        switch (LLPipeline::sRenderHighlightTextureChannel)
```

**Review by akleshchev — APPROVED** [review]
